js/src/UXCapture.js

```diff
@@ -0,0 +1,128 @@
+import ExpectedMark from "./ExpectedMark";
+import Zone from "./Zone";
+
+let _instance = null;
+
+export default class UXCapture {
+  static attachTo(window) {
+    // allow running in Node.js environment
+    if (typeof window === "undefined") {
+      window = {};
+    }
+
+    // already have UX Capture object defined, reuse it
+    if (typeof window.UX !== "undefined") {
+      return window.UX;
+    }
+
+    // assign singleton instance to window object
+    window.UX = new UXCapture(window);
+
+    _instance = window.UX;
+
+    return window.UX;
+  }
+
+  static isUserTimingSupported() {
+    return (
+      typeof _instance.window.performance !== "undefined" &&
+      typeof _instance.window.performance.mark !== "undefined" &&
+      typeof _instance.window.performance.measure !== "undefined"
+    );
+  }
+
+  static isConsoleTimeStampSupported() {
+    return (
+      typeof _instance.window.console !== "undefined" &&
+      typeof _instance.window.console.timeStamp !== "undefined"
+    );
+  }
+
+  constructor(window) {
+    this.window = window;
+
+    this.onMark = null;
+    this.onMeasure = null;
+  }
+
+  static recordMeasure(zone, lastMark) {
+    if (UXCapture.isUserTimingSupported()) {
+      // record a measure using W3C User Timing API
+      _instance.window.performance.measure(
+        zone.getLabel(),
+        "navigationStart",
```

Comment:
I'm actually wondering if we should go ahead and require that all zones declare the name of their `start` mark, and maybe default to `'navigationStart'` for now - this line would then be `zone.start.getLabel()`